packages/million/template.ts

```diff
@@ -0,0 +1,173 @@
+import { setHas$ } from './dom';
+import { AbstractBlock } from './types';
+import type { Edit, VNode } from './types';
+
+const X_CHAR = 120;
+const VOID_ELEMENTS = new Set([
+  'area',
+  'base',
+  'basefont',
+  'bgsound',
+  'br',
+  'col',
+  'command',
+  'embed',
+  'frame',
+  'hr',
+  'image',
+  'img',
+  'input',
+  'isindex',
+  'keygen',
+  'link',
+  'menuitem',
+  'meta',
+  'nextid',
+  'param',
+  'source',
+  'track',
+  'wbr',
+]);
+
+export const renderToTemplate = (
+  vnode: VNode,
+  edits: Edit[] = [],
+  path: number[] = [],
+): string => {
+  if (typeof vnode === 'string') return vnode;
+  if (
+    typeof vnode === 'number' ||
+    typeof vnode === 'bigint' ||
+    vnode === true
+  ) {
+    return String(vnode);
+  }
+  if (vnode === null || vnode === undefined || vnode === false) return '';
+
+  let props = '';
+  let children = '';
+  const current: Edit = {
+    path, // The location of the edit in in the virtual node tree
+    edits: [], // Occur on mount + patch
+    inits: [], // Occur before mount
+  };
+
+  for (let name in vnode.props) {
+    const value = vnode.props[name];
+    if (name === 'key' || name === 'ref' || name === 'children') {
+      continue;
+    }
+    if (name === 'className') name = 'class';
+    if (name === 'htmlFor') name = 'for';
+
+    if (name.startsWith('on')) {
+      const isValueHole = '__key' in value;
+      // Make objects monomorphic
+      current.edits.push({
+        type: 'event',
+        listener: isValueHole ? value.key : value,
+        name,
+        hole: isValueHole ? value.key : undefined,
+        value: undefined,
+        index: undefined,
+        patch: undefined,
+        block: undefined,
+      });
+
+      continue;
+    }
+
+    if (typeof value === 'object' && '__key' in value) {
+      current.edits.push({
+        type:
+          name === 'style'
+            ? 'style'
+            : name.charCodeAt(0) === X_CHAR
+            ? 'svg'
+            : 'attribute',
+        hole: value.__key,
+        name,
+        listener: undefined,
+        value: undefined,
+        index: undefined,
+        patch: undefined,
+        block: undefined,
+      });
+      continue;
+    }
+
+    if (value) props += ` ${name}="${String(value)}"`;
```

Comment:
This should guard against `value` being `null` (`typeof null === 'object'` is `true`, but you can't use `in` on `null`).

```suggestion
    if (value) {
      if (typeof value === 'object' && '__key' in value) {
        current.edits.push({
          type:
            name === 'style'
              ? 'style'
              : name.charCodeAt(0) === X_CHAR
              ? 'svg'
              : 'attribute',
          hole: value.__key,
          name,
          listener: undefined,
          value: undefined,
          index: undefined,
          patch: undefined,
          block: undefined,
        });
        continue;
      }
      props += ` ${name}="${String(value)}"`;
    }
```

Reply:
good catch!